[NIXL] Ignore abort on already-finished request

markmc opened the pull request:
This situation can occur when the API server receives a client disconnect (and thus sends an abort) around the same time a prefill completes and we keep the blocks (delay_free_blocks) around for a remote decode. We should assume the blocks may be used, and so we ignore the abort. If they are not used, they should be freed by the connector after a timeout.

The original error was:

```
[scheduler.py:1183] Finished sending KV transfer for request cmpl-37c560d3-5680-4bd1-97f9-7ed31a56de60-0
  File "/opt/vllm-source/vllm/v1/engine/core.py", line 292, in step
     engine_core_outputs = self.scheduler.update_from_output(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/vllm-source/vllm/v1/core/sched/scheduler.py", line 893, in update_from_output
    self._update_from_kv_xfer_finished(
  File "/opt/vllm-source/vllm/v1/core/sched/scheduler.py", line 1184, in _update_from_kv_xfer_finish>
    self._free_blocks(self.requests[req_id])
                      ~~~~~~~~~~~~~^^^^^^^^

KeyError: 'cmpl-37c560d3-5680-4bd1-97f9-7ed31a56de60-0'
```

But since #25844 we would log a warning. This fix makes it so that situation in `_update_from_kv_xfer_finish()` should never occur.

Observed under heavy load in a multi-node, llm-d 4P1D test environment. See llm-d/llm-d#187

More recently, #26012 introduced another case where this situation would cause a crash:

```
            logger.warning(
                "Releasing expired KV blocks for request %s which were "
                "retrieved by %d decode worker(s) within %d seconds.",
                req_id,
                count,
                envs.VLLM_NIXL_ABORT_REQUEST_TIMEOUT,
            )
>           self._reqs_to_process.remove(req_id)
E           KeyError: '0'

vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py:1238: KeyError
```

markmc committed:
[NIXL] Ignore abort on already-finished request

This situation can occur when the API server receives a client
disconnect (and thus sends an abort) around the same time a prefill
completes and we keep the blocks (delay_free_blocks) around for a
remote decode. We should assume the blocks may be used, and so
we ignore the abort. If they are not used, they should be freed
by the connector after a timeout.

The original error was:

```
[scheduler.py:1183] Finished sending KV transfer for request cmpl-37c560d3-5680-4bd1-97f9-7ed31a56de60-0
  File "/opt/vllm-source/vllm/v1/engine/core.py", line 292, in step
     engine_core_outputs = self.scheduler.update_from_output(
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/vllm-source/vllm/v1/core/sched/scheduler.py", line 893, in update_from_output
    self._update_from_kv_xfer_finished(
  File "/opt/vllm-source/vllm/v1/core/sched/scheduler.py", line 1184, in _update_from_kv_xfer_finish>
    self._free_blocks(self.requests[req_id])
                      ~~~~~~~~~~~~~^^^^^^^^

KeyError: 'cmpl-37c560d3-5680-4bd1-97f9-7ed31a56de60-0'
```

But since #25844 we would log a warning. This fix makes it so
that situation in `_update_from_kv_xfer_finish()` should never
occur.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented:
Related to llm-d/llm-d#187

NickLucche reviewed (changes requested):
Not sure how this wouldn't affect regular abort workflow.. 
Let's quickly discuss offline with @njhill

markmc commented:
Closing this for now - we really don't know what's going on here, my theory above doesn't make much sense, and we haven't been able to reproduce it lately.

njhill reviewed (commented):
Thanks @markmc. I spent some time studying this and I think you're right re this race condition and I can see how it could cause a double-free type key error.

markmc commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -1194,6 +1194,16 @@ def finish_requests(
+                # disconnect (and thus sends an abort) around the same time a prefill
+                # completes and we keep the blocks (delay_free_blocks) around for a
+                # remote decode. We can free the blocks now in this case.
+                logger.debug("Aborting a previously finished request %s", req_id)
+                request.status = finished_status
+                self._connector_finished(request)
+                self._free_blocks(request)
```
> I actually think it may be better here to just do nothing (i.e. just continue / skip this req).

Yeah, that's what I had this PR do in its first iteration, but that was before I understood that the decode worker could not continue in this situation, so we know we can free

> The expiration will handle blocks freeing once the timeout triggers.

I really don't love the prospect of a rare scenario whereby potentially many client-aborted requests stranded blocks for up to 8 minutes. You might kill a benchmark run, restart it, and get unpredictable results?

> We could free them immediately but that would require keeping track of another state possible state (or else we would need to ignore double-frees).

We already guard against the (impossible) possibility of another abort by checking whether the request is in `scheduler.requests`

The "but the request is already freed" check in ` _update_from_kv_xfer_finished()` is an example of us ignoring a double free already - I actually think we could make that unnecessary, but it's fine as defensive programming

Not sure I can see another issue?

> And this is only an edge case anyhow.

We do need to do something to handle this edge case though - personally I think handling the abort by freeing is likely to be less brittle than just ignoring the abort and leaving it for the expiry timer to handle

> I think for sure we shouldn't call `_connector_finished` here since that would have already been called (the reason for the race condition in the first place).

If the request is aborted, and you free the request, then the connector should also delete the timer - calling `request_finished()` is what tells the connector to do that, e.g. in NIXL:

```
	if request.status != RequestStatus.FINISHED_LENGTH_CAPPED:
            # Also include the case of a P/D Prefill request with immediate                                                                                                              
	    # block free (eg abort). Stop tracking this request.                                                                                                                         
            self._reqs_not_processed.add(request.request_id)
            return False, None

```

markmc commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -1370,7 +1380,7 @@ def _update_from_kv_xfer_finished(self, kv_connector_output: KVConnectorOutput):
             logger.debug("Finished sending KV transfer for request %s", req_id)
             if req_id not in self.requests:
-                logger.warning(
+                logger.debug(
```
IMO a warning should be actionable by the operator - if it is a warning, I think it we could expect operators to file support tickets asking "what causes this? what do I need to do to stop it happening?"

Even an info would be a bit odd - "INFO an unusual thing happened, but no need for you to worry about it"

Rob agreed it should be debug in https://github.com/vllm-project/vllm/pull/25844#discussion_r2387748219

njhill replied:
I was suggesting it to be warning if we think that it should never happen, then it's useful for it to be reported imo.

But it's expected to happen "by design" e.g. in some abort cases, then I agree debug is best and maybe there's not any need for the log at all.

For example if we changed the other logic to just ignore the abort in the case `request.is_finished()` is already true, then this should never happen AFAICT.

markmc replied:
Yes, I would prefer to get us to a point where we can just remove this

markmc committed:
[KV Transfer] Add some more detail of connector API contract

1. request_finished() should be called exactly once

2. Returning True from request_finished() means the
connector assumes responsibility for when the request
should be freed

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc committed:
[NIXL] Add assertion for abort-after-finished case

Now that we prevent against abort-after-finished, the current
assumptions in the NIXL connector are correct, but an assertion
helps document the assumption more clearly.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

njhill reviewed (approved):
Thanks @markmc! and thanks for the doc improvements!

NickLucche reviewed (approved):
Clean, thanks for the fine work @markmc
